query_database.py

```diff
@@ -0,0 +1,405 @@
+    db = {}
+    if parser.has_section(section):
+        params = parser.items(section)
+        for param in params:
+            db[param[0]] = param[1]
+
+    return db
+
+
+def insert_countries(country_entity) -> str:
+    columns: str = ''
+    values: str = ''
+
+    for key in country_entity.keys():
+        if len(columns) == 0:
+            columns += key
+        else:
+            columns += ', ' + key
+
+    for value in country_entity.values():
+        if len(values) == 0:
+            values += "'" + str(value) + "'"
+        else:
+            values += ", '" + str(value) + "'"
+
+    query = f"INSERT INTO countries ({columns}) VALUES ({values})"
+
+    return query
+
+
+def update_countries(country_entity, row) -> str:
+    column_values: str = ''
+
+    for key in country_entity:
+        column_values += f"{key} = '{country_entity[key]}', "
+
+    query = f"UPDATE countries SET {column_values[:-2]} WHERE id = {row[0]}"
+
+    return query
+
+
+def insert_countries_data(gbif_datasets, gbif_specimens, gbif_issues_flags, geocase_data):
+    """ Transforms the received countries' data to a standardised format
+        Saves the formatted data in the database by insert or update
+    """
+
+    # Temporary mapping between GBIF and GeoCASe
+    country_mapping = {
+        'DE': 'Germany',
+        'GB': 'UK',
+        'EE': 'Estonia',
+        'AT': 'Austria',
+        'FI': 'Finland',
+        'NL': 'The Netherlands'
+    }
+
+    # For each organisation, prepare and save data to database
+    for country in gbif_datasets['countries']:
+        # Check if GeoCASe data from country is present
+        geocase_dummy: dict = {}
+
+        if country in country_mapping:
+            geocase_dummy = geocase_data['countries'][country_mapping[country]]
+
+        # Initiate basic entity for database
+        country_entity = {
+            'country_code': country,
+            'country_name': 'test',
+            'datasets_count': json.dumps({
+                'gbif': gbif_datasets['countries'][country]
+            }),
+            'specimens_count': json.dumps({
+                'gbif': gbif_specimens['countries'][country],
+                'geocase': geocase_dummy
+            }),
+            'issues_flags': json.dumps(gbif_issues_flags['countries'][country])
+        }
+
+        # Create database connection
+        params = config()
+        conn = psycopg2.connect(**params)
+        cur = conn.cursor()
+
+        # Check if entity already exists (for same month)
+        query = f"SELECT * FROM countries WHERE EXTRACT(MONTH from datetime) = {current_num_month} AND country_code = '{country}'"
+        cur.execute(query)
+        row = cur.fetchone()
+
+        if row:
+            # Update existing record
+            query = update_countries(country_entity, row)
+        else:
+            # Insert new record
+            query = insert_countries(country_entity)
+
+        # Execute query
```

Comment:
See if you can use the try-with-resources functionality then you don't need to explicitly commit and close the connection.
For sql alchemy:
https://docs.sqlalchemy.org/en/14/core/connections.html

Reply:
Replaced plain Psycopg2 database queries with SQL Alchemy queries that use the 'with' method